Userland/Libraries/LibGfx/ImageFormats/JPEG2000Loader.cpp

```diff
@@ -0,0 +1,172 @@
+[[maybe_unused]] static constexpr u8 marker_id_string[] = { 0xFF, 0x4F, 0xFF, 0x51 };
+
+// This is the marker for the box version.
+// T.800 Annex I, JP2 file format syntax, I.5.1 JPEG 2000 Signature box
+static constexpr u8 jp2_id_string[] = { 0x00, 0x00, 0x00, 0x0C, 0x6A, 0x50, 0x20, 0x20, 0x0D, 0x0A, 0x87, 0x0A };
+
+struct JPEG2000LoadingContext {
+    enum class State {
+        NotDecoded = 0,
+        Error,
+    };
+    State state { State::NotDecoded };
+    ReadonlyBytes codestream_data;
+    Optional<ReadonlyBytes> icc_data;
+
+    IntSize size;
+
+    ISOBMFF::BoxList boxes;
+};
+
+static ErrorOr<void> decode_jpeg2000_header(JPEG2000LoadingContext& context, ReadonlyBytes data)
+{
+    if (!JPEG2000ImageDecoderPlugin::sniff(data))
+        return Error::from_string_literal("JBIG2LoadingContext: Invalid JBIG2 header");
+
+    auto reader = TRY(Gfx::ISOBMFF::Reader::create(TRY(try_make<FixedMemoryStream>(data))));
+    context.boxes = TRY(reader.read_entire_file());
+
+    // I.2.2 File organization
+    // "A particular order of those boxes in the file is not generally implied. However, the JPEG 2000 Signature box
+    //  shall be the first box in a JP2 file, the File Type box shall immediately follow the JPEG 2000 Signature box
+    //  and the JP2 Header box shall fall before the Contiguous Codestream box."
+    if (context.boxes.size() < 4)
+        return Error::from_string_literal("JPEG2000ImageDecoderPlugin: Expected at least four boxes");
+
+    // Required toplevel boxes: signature box, file type box, jp2 header box, contiguous codestream box.
+
+    if (context.boxes[0]->box_type() != ISOBMFF::BoxType::JPEG2000SignatureBox)
+        return Error::from_string_literal("JPEG2000ImageDecoderPlugin: Expected JPEG2000SignatureBox as first box");
+    if (context.boxes[1]->box_type() != ISOBMFF::BoxType::FileTypeBox)
+        return Error::from_string_literal("JPEG2000ImageDecoderPlugin: Expected FileTypeBox as first box");
+
+    Optional<size_t> jp2_header_box_index;
+    Optional<size_t> contiguous_codestream_box_index;
+    for (size_t i = 2; i < context.boxes.size(); ++i) {
+        if (context.boxes[i]->box_type() == ISOBMFF::BoxType::JPEG2000HeaderBox) {
+            // "Within a JP2 file, there shall be one and only one JP2 Header box."
+            if (jp2_header_box_index.has_value())
+                return Error::from_string_literal("JPEG2000ImageDecoderPlugin: Multiple JP2 Header boxes");
+            jp2_header_box_index = i;
+        }
+        if (context.boxes[i]->box_type() == ISOBMFF::BoxType::JPEG2000ContiguousCodestreamBox && !contiguous_codestream_box_index.has_value()) {
+            // "a conforming reader shall ignore all codestreams after the first codestream found in the file.
+            //  Contiguous Codestream boxes may be found anywhere in the file except before the JP2 Header box."
+            contiguous_codestream_box_index = i;
```

Comment:
I definitely stumbled a bit while looking for the "except before the JP2 Header box" part, not noticing the index compare below. I feel like we might as well stop here as early as possible. Any followup codestream boxes aren't relevant to us anyways.

Reply:
Good comment, thanks. Done.